File: ResourceManager/Models/DbContextExtensions.cs
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ResourceManager.Models
{
    internal static class DbContextExtensions
    {
        public static void UpdateManyToMany<T, TKey>(
            this DbContext db,
            IEnumerable<T> dbEntries,
            IEnumerable<T> updatedEntries,
            Func<T, TKey> keyRetrievalFunction)
            where T : class
        {
            var oldItems = dbEntries.ToList();
            var newItems = updatedEntries.ToList();
            var toBeRemoved = oldItems.LeftComplementRight(newItems, keyRetrievalFunction);
            var toBeAdded = newItems.LeftComplementRight(oldItems, keyRetrievalFunction);
            var toBeUpdated = oldItems.Intersect(newItems, keyRetrievalFunction);

            db.Set<T>().RemoveRange(toBeRemoved);
            db.Set<T>().AddRange(toBeAdded);
            foreach (var entity in toBeUpdated)
            {
                var changed = newItems.Single(i => keyRetrievalFunction.Invoke(i)!.Equals(keyRetrievalFunction.Invoke(entity)));
                db.Entry(entity).CurrentValues.SetValues(changed);
            }
        }
        public static IEnumerable<T> LeftComplementRight<T, TKey>(
            this IEnumerable<T> left,
            IEnumerable<T> right,
            Func<T, TKey> keyRetrievalFunction)
        {
            var leftSet = left.ToList();
            var rightSet = right.ToList();

            var leftSetKeys = leftSet.Select(keyRetrievalFunction);
            var rightSetKeys = rightSet.Select(keyRetrievalFunction);

            var deltaKeys = leftSetKeys.Except(rightSetKeys);
            var leftComplementRightSet = leftSet.Where(i => deltaKeys.Contains(keyRetrievalFunction.Invoke(i)));
            
            return leftComplementRightSet;
        }

        public static IEnumerable<T> Intersect<T, TKey>(
            this IEnumerable<T> left,
            IEnumerable<T> right,
            Func<T, TKey> keyRetrievalFunction)
        {
            var leftSet = left.ToList();
            var rightSet = right.ToList();

            var leftSetKeys = leftSet.Select(keyRetrievalFunction);
            var rightSetKeys = rightSet.Select(keyRetrievalFunction);

            var intersectKeys = leftSetKeys.Intersect(rightSetKeys);
            var intersectionEntities = leftSet.Where(i => intersectKeys.Contains(keyRetrievalFunction.Invoke(i)));
            
            return intersectionEntities;
        }
    }
}

File: ResourceManager/Models/DbContextExtensions.vb
Imports Microsoft.EntityFrameworkCore
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Runtime.CompilerServices

Namespace ResourceManager.Models
    Friend Module DbContextExtensions
        <Extension()>
        Public Sub UpdateManyToMany(Of T As Class, TKey)(ByVal db As DbContext, ByVal dbEntries As IEnumerable(Of T), ByVal updatedEntries As IEnumerable(Of T), ByVal keyRetrievalFunction As Func(Of T, TKey))
            Dim oldItems = dbEntries.ToList()
            Dim newItems = updatedEntries.ToList()
            Dim toBeRemoved = oldItems.LeftComplementRight(newItems, keyRetrievalFunction)
            Dim toBeAdded = newItems.LeftComplementRight(oldItems, keyRetrievalFunction)
            Dim toBeUpdated = oldItems.Intersect(newItems, keyRetrievalFunction)

            db.Set(Of T)().RemoveRange(toBeRemoved)
            db.Set(Of T)().AddRange(toBeAdded)
            For Each entity In toBeUpdated
                Dim changed = newItems.[Single](Function(i) keyRetrievalFunction.Invoke(i).Equals(keyRetrievalFunction.Invoke(entity)))
                db.Entry(entity).CurrentValues.SetValues(changed)
            Next
        End Sub
        <Extension()>
        Public Function LeftComplementRight(Of T, TKey)(ByVal left As IEnumerable(Of T), ByVal right As IEnumerable(Of T), ByVal keyRetrievalFunction As Func(Of T, TKey)) As IEnumerable(Of T)
            Dim leftSet = left.ToList()
            Dim rightSet = right.ToList()

            Dim leftSetKeys = leftSet.[Select](keyRetrievalFunction)
            Dim rightSetKeys = rightSet.[Select](keyRetrievalFunction)

            Dim deltaKeys = leftSetKeys.Except(rightSetKeys)
            Dim leftComplementRightSet = leftSet.Where(Function(i) deltaKeys.Contains(keyRetrievalFunction.Invoke(i)))

            Return leftComplementRightSet
        End Function

        <Extension()>
        Public Function Intersect(Of T, TKey)(ByVal left As IEnumerable(Of T), ByVal right As IEnumerable(Of T), ByVal keyRetrievalFunction As Func(Of T, TKey)) As IEnumerable(Of T)
            Dim leftSet = left.ToList()
            Dim rightSet = right.ToList()

            Dim leftSetKeys = leftSet.[Select](keyRetrievalFunction)
            Dim rightSetKeys = rightSet.[Select](keyRetrievalFunction)

            Dim intersectKeys = leftSetKeys.Intersect(rightSetKeys)
            Dim intersectionEntities = leftSet.Where(Function(i) intersectKeys.Contains(keyRetrievalFunction.Invoke(i)))

            Return intersectionEntities
        End Function
    End Module
End Namespace
